File: 백준/Gold/1749. 점수따먹기/점수따먹기.py
import sys
input = sys.stdin.readline
N, M = map(int, input().split()) # N 배열의 크기, M 연산 횟수
arr = [[0]*(M+1)]+ [[0] + list(map(int, input().split())) for _ in range(N)] # 배열의 크기 + 1 만큼 생성, 0은 패딩
prefix = [[0 for _ in range(M+1)] for _ in range(N+1)] # 누적합
for i in range(1,N+1): # 2차원 배열 누적합
    for j in range(1,M+1):
        prefix[i][j] = prefix[i][j-1] + prefix[i-1][j] + arr[i][j] - prefix[i-1][j-1]
ans = -1e9
for x1 in range(1,N+1):
    for y1 in range(1, M+1):
        for x2 in range(x1, N+1):
            for y2 in range(y1, M+1):
                ans = max(ans, prefix[x2][y2]-prefix[x1-1][y2]-prefix[x2][y1-1]+prefix[x1-1][y1-1])

print(ans)

File: 백준/Gold/1749. 점수따먹기/점수따먹기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Main {
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st =new StringTokenizer(br.readLine());
		int N = Integer.parseInt(st.nextToken());
		int M = Integer.parseInt(st.nextToken());

		int[][] grid = new int[N][M];
		for (int i=0; i<N; i++) {
			grid[i] = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
		}
		int ans = Integer.MIN_VALUE;

		int[][] prefix = new int[N+1][M+1];
		for (int i=0; i<N; i++) {
			for (int j=0; j<M; j++) {
				prefix[i + 1][j + 1] = grid[i][j] + prefix[i + 1][j] + prefix[i][j+1] - prefix[i][j];
			}
		}

		for (int i=1; i<=N; i++) {
			for (int j=1; j<=M; j++) {
				for (int h=i; h<=N; h++) {
					for (int k=j; k<=M; k++) {
						ans = Math.max(ans, prefix[h][k] - prefix[i-1][k] - prefix[h][j-1] + prefix[i-1][j-1]);
					}
				}
			}
		}

		System.out.print(ans);
	}
}
